src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -840,93 +867,889 @@ class ApplyHelper
+                mDiagnosticEvents.pushError(
+                    SCE_BUDGET, SCEC_EXCEEDED_LIMIT,
+                    "total events size exceeds network config maximum",
+                    {makeU64SCVal(mMetrics.mEmitEventByte),
+                     makeU64SCVal(
+                         mSorobanConfig.txMaxContractEventsSizeBytes())});
+                mOpFrame.innerResult(mRes).code(
+                    INVOKE_HOST_FUNCTION_RESOURCE_LIMIT_EXCEEDED);
+                return {false, {}};
+            }
+            ContractEvent evt;
+            xdr::xdr_from_opaque(buf.data, evt);
+            success.events.emplace_back(evt);
+        }
+
+        mOpFrame.maybePopulateDiagnosticEvents(mAppConfig, out, mMetrics,
+                                               mDiagnosticEvents);
+
+        mMetrics.mEmitEventByte +=
+            static_cast<uint32>(out.result_value.data.size());
+        if (mSorobanConfig.txMaxContractEventsSizeBytes() <
+            mMetrics.mEmitEventByte)
+        {
+            mDiagnosticEvents.pushError(
+                SCE_BUDGET, SCEC_EXCEEDED_LIMIT,
+                "return value pushes events size above network config maximum",
+                {makeU64SCVal(mMetrics.mEmitEventByte),
+                 makeU64SCVal(mSorobanConfig.txMaxContractEventsSizeBytes())});
+            mOpFrame.innerResult(mRes).code(
+                INVOKE_HOST_FUNCTION_RESOURCE_LIMIT_EXCEEDED);
+            return {false, {}};
+        }
+
+        if (!mRefundableFeeTracker->consumeRefundableSorobanResources(
+                mMetrics.mEmitEventByte, out.rent_fee,
+                mLedgerInfo.getLedgerVersion(), mSorobanConfig, mAppConfig,
+                mOpFrame.mParentTx, mDiagnosticEvents))
+        {
+            mOpFrame.innerResult(mRes).code(
+                INVOKE_HOST_FUNCTION_INSUFFICIENT_REFUNDABLE_FEE);
+            return {false, {}};
+        }
+
+        xdr::xdr_from_opaque(out.result_value.data, success.returnValue);
+        mOpFrame.innerResult(mRes).code(INVOKE_HOST_FUNCTION_SUCCESS);
+        mOpFrame.innerResult(mRes).success() = xdrSha256(success);
+
+        mOpMeta.getEventManager().setEvents(std::move(success.events));
+        mOpMeta.setSorobanReturnValue(success.returnValue);
+        mMetrics.mSuccess = true;
+
+        return {true, std::move(opEntryMap), std::move(restoredKeys)};
+    }
+};
+
+InvokeHostFunctionOpFrame::InvokeHostFunctionOpFrame(
+    Operation const& op, TransactionFrame const& parentTx)
+    : OperationFrame(op, parentTx)
+    , mInvokeHostFunction(mOperation.body.invokeHostFunctionOp())
+{
+}
+
+bool
+InvokeHostFunctionOpFrame::isOpSupported(LedgerHeader const& header) const
+{
+    return header.ledgerVersion >= 20;
+}
+
+void
+InvokeHostFunctionOpFrame::maybePopulateDiagnosticEvents(
+    Config const& cfg, InvokeHostFunctionOutput const& output,
+    HostFunctionMetrics const& metrics, DiagnosticEventManager& buffer) const
+{
+    if (cfg.ENABLE_SOROBAN_DIAGNOSTIC_EVENTS)
```

Comment:
nit: Maybe return early to reduce nesting?

Reply:
done